python/lsst/pipe/base/graph/_loadHelpers.py

```diff
@@ -0,0 +1,310 @@
+            # only some bytes are being read using the reader specialized for
+            # this class
+            nodes = list(nodes)
+            _readBytes = self._readBytes
+        # create a container for loaded data
+        quanta: DefaultDict[TaskDef, Set[Quantum]] = defaultdict(set)
+        quantumToNodeId: Dict[Quantum, NodeId] = {}
+        loadedTaskDef = {}
+        # loop over the nodes specified above
+        for node in nodes:
+            # Get the bytes to read from the map
+            start, stop = self.map[node]
+            start += self.headerSize
+            stop += self.headerSize
+
+            # read the specified bytes, will be overloaded by subclasses
+            # bytes are compressed, so decompress them
+            dump = lzma.decompress(_readBytes(start, stop))
+
+            # reconstruct node
+            qNode = pickle.loads(dump)
+
+            # read the saved node, name. If it has been loaded, attach it, if
+            # not read in the taskDef first, and then load it
+            nodeTask = qNode.taskDef
+            if nodeTask not in loadedTaskDef:
+                # Get the byte ranges corresponding to this taskDef
+                start, stop = self.taskDefMap[nodeTask]
+                start += self.headerSize
+                stop += self.headerSize
+
+                # load the taskDef, this method call will be overloaded by
+                # subclasses.
+                # bytes are compressed, so decompress them
+                taskDef = pickle.loads(lzma.decompress(_readBytes(start, stop)))
+                loadedTaskDef[nodeTask] = taskDef
+            # Explicitly overload the "frozen-ness" of nodes to attach the
+            # taskDef back into the un-persisted node
+            object.__setattr__(qNode, 'taskDef', loadedTaskDef[nodeTask])
+            quanta[qNode.taskDef].add(qNode.quantum)
+
+            # record the node for later processing
+            quantumToNodeId[qNode.quantum] = qNode.nodeId
+
+        # Get the id of this graph
+        _buildId = qNode.nodeId.buildId if nodes else None  # type: ignore
+
+        # construct an empty new QuantumGraph object, and run the associated
+        # creation method with the un-persisted data
+        qGraph = object.__new__(QuantumGraph)
+        qGraph._buildGraphs(quanta, _quantumToNodeId=quantumToNodeId, _buildId=_buildId)
+        return qGraph
+
+    def _readBytes(self, start: int, stop: int) -> bytes:
+        """Loads the specified byte range from the ButlerURI object
+
+        In the base class, this actually will read all the bytes into a buffer
+        from the specified ButlerURI object. Then for each method call will
+        return the requested byte range. This is the most flexible
+        implementation, as no special read is required. This will not give a
+        speed up with any sub graph reads though.
+        """
+        if not hasattr(self, 'buffer'):
+            self.buffer = self.uriObject.read()
```

Comment:
I"m not sure `hasattr` is justified in this case, could you instead set `self.buffer = None` in constructor and just check for `is None`?

Reply:
I did this because each subclass is going to be doing its own thing, and I didnt really want variables that were not important to the current class in the to be in the current class. I can think of a few other ways to do it, but I am not sure what you dislike about this approach, it is writing out exactly what I am trying to ask, and is not really much overhead compared to a None check.

Reply:
Having "dynamic" attributes that can be added or removed is a bit too much dynamic to my taste, probably due to my C++ background. If subclasses are doing something very different then you probably want a different base class or interface for them? This code sort of assumes that subclasses do not add "buffer" attribute or if they do add it then they know what they do. This sort of dependency may be better expressed as some (abstract) method in this class.